sitemedia/js/controllers/iiif_controller.js

```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
-        while (this.iiifContainerTarget.firstChild) {
-            this.iiifContainerTarget.removeChild(
-                this.iiifContainerTarget.firstChild
+        const zoomSlider = element.querySelector("input[type='range']");
+        const zoomSliderLabel = element.querySelector(
+            "label[for^='zoom-slider']"
+        );
+        const zoomToggle = element.querySelector(
+            "input[type='checkbox'][id^='zoom-toggle']"
+        );
+        const image = element.querySelector("img.iiif-image");
+        viewer.addHandler("open", () => {
+            // ensure image is positioned in top-left corner of viewer
+            this.resetBounds(viewer);
+            if (isMobile) {
+                // zoom to 110% if on mobile
+                viewer.viewport.zoomTo(1.1);
+            }
+            // initialize zoom slider
+            zoomSlider.setAttribute("min", minZoom);
+            // use toPrecision to ensure no extra pixels on the right of the slider
+            zoomSlider.setAttribute(
+                "max",
+                viewer.viewport.getMaxZoom().toPrecision(2)
             );
+            zoomSlider.addEventListener("input", (evt) => {
+                // Handle changes in the zoom slider
+                let zoom = parseFloat(evt.currentTarget.value);
+                let deactivating = false;
+                if (zoom <= minZoom) {
+                    // When zoomed back out to 100%, deactivate OSD
+                    zoom = minZoom;
+                    this.resetBounds(viewer);
+                    this.deactivateDeepZoom(element, image);
+                    evt.currentTarget.value = minZoom;
+                    deactivating = true;
+                } else {
+                    // Zoom to the chosen percentage
+                    viewer.viewport.zoomTo(zoom);
+                }
+                this.updateZoomUI(
+                    zoom,
+                    deactivating,
+                    zoomSlider,
+                    zoomSliderLabel
+                );
+            });
+            // initialize mobile zoom toggle
+            zoomToggle.addEventListener("input", (evt) => {
+                // always first zoom to 100%
+                viewer.viewport.zoomTo(1.0);
+                if (!evt.currentTarget.checked) {
+                    // wait to reset bounds until element is hidden
+                    setTimeout(() => this.resetBounds(viewer), 300);
+                    this.deactivateDeepZoom(element, image);
+                } else {
+                    // reset bounds immediately and zoom to 110%
+                    this.resetBounds(viewer);
+                    viewer.viewport.zoomTo(1.1);
+                }
+            });
+        });
+        viewer.addHandler("zoom", (evt) => {
+            // Handle changes in the canvas zoom
+            let { zoom } = evt;
+            if (zoom <= minZoom) {
```

Comment:
how is it possible to zoom to less than `minZoom`?

Reply:
You would think it wouldn't be, but it is! Using the scroll wheel on my mouse, I get these two `zoom` events in rapid succession:

```
zoomed to 0.8333333333333334
minZoom: 1

zoomed to 1
minZoom: 1
```

Reply:
ah, weird. I would have thought that the minzoom config in the OSD seadragon would prevent that.